discovery_engine_core/web-api/src/handlers.rs

```diff
@@ -31,11 +44,61 @@ pub(crate) async fn handle_ranked_documents(
         .await
         .map_err(handle_user_state_op_error)?;
 
-    let mut documents = db.documents.clone();
+    if user_interests.positive.is_empty() {
+        return Err(warp::reject::custom(NoCenterOfInterestsError));
+    }
+
+    let max_cois = MAX_COIS.min(user_interests.positive.len());
+    let cois = &user_interests.positive[0..max_cois];
+    let horizon = db.coi.config().horizon();
+    let coi_weights = compute_coi_weights(cois, horizon);
+    let document_futures = cois
+        .iter()
+        .zip(&coi_weights)
+        .map(|(coi, weight)| async {
+            #[allow(
+                // fine as max documents count is small enough
+                clippy::cast_precision_loss,
+                // fine as weight should be between 0 and 1
+                clippy::cast_sign_loss,
+                // fine as number of neighbors is small enough
+                clippy::cast_possible_truncation
+            )]
+            let k_neighbors = (*weight * MAX_DOCUMENTS_COUNT as f32).ceil() as usize;
 
-    // TODO TO-3339: Return 500 with the correct kind if this fail
-    let scores = db.coi.score(&documents, &user_interests).unwrap();
-    rank(&mut documents, &scores);
+            db.elastic_state
+                .get_documents_by_embedding(&coi.point, k_neighbors, MAX_DOCUMENTS_COUNT)
+                .await
+        })
+        .collect_vec();
```

Comment:
:art: instead of collecting them into a `Vec<Future>` and call `join_all()`, we could also collect them into a `FutureUnordered` and await them. this can provide some performance benefits over the former which uses `FuturesOrdered` internally. if you like you can try to rewrite this, but it's probably not crucial right now.

Reply:
this will be handled by https://xainag.atlassian.net/browse/TO-3376